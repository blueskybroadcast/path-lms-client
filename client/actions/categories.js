import normalize from 'jsonapi-normalizer';

import {
  FETCH_CATEGORIES_SUCCESS, DELETE_CATEGORY_SUCCESS, FETCH_CATEGORIES_REQUEST,
  ADD_CATEGORY_REQUEST, ADD_CATEGORY_SUCCESS, ADD_CATEGORY_FAILURE
} from './types';

export const fetchCategories = () => async (dispatch, getState, api) => {
  dispatch({ type: FETCH_CATEGORIES_REQUEST });
  const res = await api.get('/categories');

  dispatch({
    type: FETCH_CATEGORIES_SUCCESS,
    payload: normalize(res.data)
  });
};

export const addCategory = category => async (dispatch, getState, api) => {
  dispatch({ type: ADD_CATEGORY_REQUEST });

  const { name, parentId, hidden } = category;

  await api.post('/categories', {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch({ type: ADD_CATEGORY_SUCCESS });
    dispatch(fetchCategories());
  }).catch(() => {
    dispatch({ type: ADD_CATEGORY_FAILURE });
  });
};

export const deleteCategory = id => async (dispatch, getState, api) => {
  await api.delete(`/categories/${id}`).then(() => {
    dispatch({
      type: DELETE_CATEGORY_SUCCESS,
      payload: id
    });
  });
};

export const editCategory = (id, category) => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.put(`/categories/${id}`, {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch(fetchCategories());
  });
};
