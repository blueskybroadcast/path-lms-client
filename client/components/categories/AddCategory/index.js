import AddCategory from './container';

export default AddCategory;
