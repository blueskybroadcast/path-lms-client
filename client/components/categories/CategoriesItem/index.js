import CategoriesItem from './container';

export default CategoriesItem;
