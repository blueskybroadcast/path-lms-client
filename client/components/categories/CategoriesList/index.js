import CategoriesList from './container';

export default CategoriesList;
