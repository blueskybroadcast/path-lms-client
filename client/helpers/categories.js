export const getParentListLevelOne = (categories) => {
  const parentsLevelOne = categories.filter(cat => cat.level === 0 && cat.childCount > 0);
  return parentsLevelOne.map(({ name, id }) => ({
    name,
    list: categories.filter(cat => cat.parentId === id)
  }));
};

export const getSortedCategories = (categoriesIds, categoriesData) => {
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        level: 0,
        name: data.name,
        childCount: 0,
        hidden: data.hidden
      });
    }
  });

  const nextCategories = [...categories];
  let nextAddedCategoriesCount = 0;

  categories.forEach((cat, index) => {
    categoriesIds.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        nextCategories.splice(
          index + 1 + nextAddedCategoriesCount,
          0,
          {
            id,
            level: 1,
            name: data.name,
            childCount: 0,
            parentId,
            hidden: data.hidden
          }
        );
        categories[index].childCount += 1;
        nextAddedCategoriesCount += 1;
      }
    });
  });

  const finalCategories = [...nextCategories];
  let finalAddedCategoriesCount = 0;

  nextCategories.forEach((cat, index) => {
    if (cat.level === 1) {
      categoriesIds.forEach((id) => {
        const data = categoriesData[id];
        if (cat.id === data.parent.id) {
          finalCategories.splice(
            index + 1 + finalAddedCategoriesCount,
            0,
            {
              id,
              level: 2,
              name: data.name,
              hidden: data.hidden
            }
          );
          nextCategories[index].childCount += 1;
          finalAddedCategoriesCount += 1;
        }
      });
    }
  });

  return finalCategories;
};

export const getCategoriesForSelect = (categoriesIds, categoriesData) => {
  if (!categoriesIds.length) return [];
  const remainingCategories = [...categoriesIds];
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        name: data.name,
        hidden: data.hidden,
        children: []
      });
      const itemToRemoveIndex = remainingCategories.indexOf(id);
      remainingCategories.splice(itemToRemoveIndex, 1);
    }
  });

  const finalCategories = [...categories];

  categories.forEach((cat, index) => {
    remainingCategories.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        const levelOneCat = {
          id,
          name: data.name,
          hidden: data.hidden,
          children: []
        };

        remainingCategories.forEach((childId) => {
          const childData = categoriesData[childId];
          const childParentId = childData.parent.id;

          if (id === childParentId) {
            const levelTwoCat = {
              id: childId,
              name: childData.name,
              hidden: childData.hidden,
              children: []
            };

            levelOneCat.children.push(levelTwoCat);
          }
        });

        finalCategories[index].children.push(levelOneCat);
      }
    });
  });

  return finalCategories;
};
