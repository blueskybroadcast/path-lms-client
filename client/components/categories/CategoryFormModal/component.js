import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Modal, Form } from 'react-bootstrap';

import { getParentListLevelOne } from '../../../helpers/categories';

const initialState = {
  name: '',
  parentId: '',
  hidden: false
};

class CategoryFormModal extends React.Component {
  state = initialState

  componentDidMount() {
    this.initialLoad();
  }

  initialLoad = () => {
    const {
      name, parentId, hidden
    } = this.props;
    this.setState({
      name,
      parentId,
      hidden
    });
  }

  handleClose = () => {
    const { handleClose } = this.props;
    handleClose();
    this.initialLoad();
  }

  handleChange = ({ target }) => {
    const {
      type, checked, value, name
    } = target;

    const val = type === 'checkbox' ? checked : value;

    this.setState({
      [name]: val
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    if (!this.validateForm()) {
      const { handleSubmit } = this.props;
      const { name, parentId, hidden } = this.state;
      handleSubmit({ name, parentId, hidden });
      this.setState(initialState);
      this.handleClose();
    }
  }

  validateForm = () => {
    const { name } = this.state;
    const isInvalid = !name;
    return isInvalid;
  }

  render() {
    const { name, parentId, hidden } = this.state;
    const {
      show, mode, categoriesSortedList, submitting
    } = this.props;
    const isFormInvalid = this.validateForm();

    const parentListLevelZero = categoriesSortedList.filter(cat => cat.level === 0);
    const parentListLevelOne = getParentListLevelOne(categoriesSortedList);
    const submitButtonLabel = mode === 'create' ? 'Create Category' : 'Update Category';

    return (
      <Modal
        show={show}
        onHide={this.handleClose}
        size="lg"
      >
        <div className="modal-inner">
          <section className="modal-form">
            <header>
              <h4>{mode === 'create' ? 'Add Category' : 'Edit Category'}</h4>
            </header>
            <div className="form-wrapper">
              <form
                noValidate="novalidate"
                className="simple_form edit_category"
                onSubmit={this.handleSubmit}
              >
                <div className="form-inner">
                  <fieldset>
                    <div className="control-group string required category_name">
                      <label
                        className="string required control-label"
                        htmlFor="category_name"
                      >
                        Name
                        <abbr title="required" className="required">*</abbr>
                      </label>
                      <div className="controls">
                        <input
                          className="string required"
                          id="category_name"
                          required="required"
                          aria-required="true"
                          type="text"
                          onChange={this.handleChange}
                          value={name}
                          name="name"
                        />
                      </div>
                    </div>
                  </fieldset>
                  <fieldset>
                    <div className="control-group integer optional category_parent_id">
                      <label
                        className="integer optional control-label"
                        htmlFor="parentId"
                      >
                        Parent
                      </label>
                      <div className="controls">
                        <Form.Control
                          as="select"
                          name="parentId"
                          onChange={this.handleChange}
                          value={parentId}
                        >
                          <option value="">-------------</option>
                          <optgroup label="Top Level Categories">
                            {parentListLevelZero.map(cat => (
                              <option
                                key={cat.id}
                                value={cat.id}
                              >
                                {cat.name}
                              </option>
                            ))}
                          </optgroup>
                          {parentListLevelOne.map(parent => (
                            <optgroup
                              label={parent.name}
                              key={parent.name}
                            >
                              {parent.list.map(category => (
                                <option
                                  value={category.id}
                                  key={category.id}
                                >
                                  {category.name}
                                </option>
                              ))}
                            </optgroup>
                          ))}
                        </Form.Control>
                      </div>
                    </div>
                  </fieldset>
                  <fieldset>
                    <div className="control-group active-toggle">
                      <label htmlFor="category_hidden">
                        Hidden
                      </label>
                      <div className="controls">
                        <label
                          className="label-switch"
                          role="checkbox"
                          tabIndex="0"
                          htmlFor="category_hidden"
                          aria-checked="false"
                        >
                          <input
                            className="boolean optional switcher optional"
                            id="category_hidden"
                            type="checkbox"
                            checked={hidden}
                            onChange={this.handleChange}
                            name="hidden"
                          />
                          <div className="checkbox" />
                        </label>
                        <p className="help-block">
                          Hidden categories only display to admins. These also display in reports
                        </p>
                      </div>
                    </div>
                  </fieldset>
                </div>
                <footer className="confirm">
                  <button
                    type="submit"
                    className={classNames({
                      btn: true,
                      disabled: isFormInvalid || submitting,
                      submitting
                    })}
                    onClick={this.handleSubmit}
                    disabled={isFormInvalid || submitting}
                  >
                    { submitting
                      ? 'Processing...'
                      : submitButtonLabel
                    }
                  </button>

                  <button
                    className="cancel close-modal"
                    type="button"
                    onClick={this.handleClose}
                  >
                    Cancel
                  </button>
                </footer>
              </form>
            </div>
          </section>
        </div>
      </Modal>
    );
  }
}

CategoryFormModal.propTypes = {
  name: PropTypes.string,
  parentId: PropTypes.string,
  hidden: PropTypes.bool,
  mode: PropTypes.string,
  handleSubmit: PropTypes.func.isRequired,
  handleClose: PropTypes.func.isRequired,
  show: PropTypes.bool.isRequired,
  submitting: PropTypes.bool.isRequired,
  categoriesSortedList: PropTypes.arrayOf(PropTypes.object).isRequired
};

CategoryFormModal.defaultProps = {
  name: '',
  hidden: false,
  parentId: '',
  mode: 'create'
};

export default CategoryFormModal;
