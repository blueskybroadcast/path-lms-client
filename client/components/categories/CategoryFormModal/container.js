import { connect } from 'react-redux';
import CategoryFormModal from './component';

import { categoriesSortedIdsSelector } from '../../../selectors/categoriesSelectors';

const mapStateToProps = state => ({
  categoriesSortedList: categoriesSortedIdsSelector(state)
});

export default connect(mapStateToProps)(CategoryFormModal);
