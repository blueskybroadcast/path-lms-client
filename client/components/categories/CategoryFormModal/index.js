import CategoryFormModal from './container';

export default CategoryFormModal;
