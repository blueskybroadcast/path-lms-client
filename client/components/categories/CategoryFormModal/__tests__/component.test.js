import React from 'react';
import { shallow } from 'enzyme';
import CategoryFormModal from '../component';

import { categoriesSortedList } from '../../../../fixtures/categories';

test('should render CategoryFormModal correctly', () => {
  const wrapper = shallow(
    <CategoryFormModal
      handleSubmit={jest.fn()}
      handleClose={jest.fn()}
      show
      submitting={false}
      categoriesSortedList={categoriesSortedList}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
