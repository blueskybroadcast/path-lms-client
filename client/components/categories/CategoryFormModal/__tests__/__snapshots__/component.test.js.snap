// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CategoryFormModal correctly 1`] = `
<ForwardRef(Bootstrap(Modal))
  onHide={[Function]}
  show={true}
  size="lg"
>
  <div
    className="modal-inner"
  >
    <section
      className="modal-form"
    >
      <header>
        <h4>
          Add Category
        </h4>
      </header>
      <div
        className="form-wrapper"
      >
        <form
          className="simple_form edit_category"
          noValidate="novalidate"
          onSubmit={[Function]}
        >
          <div
            className="form-inner"
          >
            <fieldset>
              <div
                className="control-group string required category_name"
              >
                <label
                  className="string required control-label"
                  htmlFor="category_name"
                >
                  Name
                  <abbr
                    className="required"
                    title="required"
                  >
                    *
                  </abbr>
                </label>
                <div
                  className="controls"
                >
                  <input
                    aria-required="true"
                    className="string required"
                    id="category_name"
                    name="name"
                    onChange={[Function]}
                    required="required"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </fieldset>
            <fieldset>
              <div
                className="control-group integer optional category_parent_id"
              >
                <label
                  className="integer optional control-label"
                  htmlFor="parentId"
                >
                  Parent
                </label>
                <div
                  className="controls"
                >
                  <ForwardRef(ContextTransform)
                    as="select"
                    name="parentId"
                    onChange={[Function]}
                    value=""
                  >
                    <option
                      value=""
                    >
                      -------------
                    </option>
                    <optgroup
                      label="Top Level Categories"
                    >
                      <option
                        key="28"
                        value="28"
                      >
                        Another Level 0
                      </option>
                      <option
                        key="25"
                        value="25"
                      >
                        Level 0
                      </option>
                      <option
                        key="29"
                        value="29"
                      >
                        Super Level 0
                      </option>
                    </optgroup>
                    <optgroup
                      key="Another Level 0"
                      label="Another Level 0"
                    >
                      <option
                        key="30"
                        value="30"
                      >
                        Another Level 0 child
                      </option>
                    </optgroup>
                    <optgroup
                      key="Level 0"
                      label="Level 0"
                    >
                      <option
                        key="26"
                        value="26"
                      >
                        Level 0 child
                      </option>
                    </optgroup>
                  </ForwardRef(ContextTransform)>
                </div>
              </div>
            </fieldset>
            <fieldset>
              <div
                className="control-group active-toggle"
              >
                <label
                  htmlFor="category_hidden"
                >
                  Hidden
                </label>
                <div
                  className="controls"
                >
                  <label
                    aria-checked="false"
                    className="label-switch"
                    htmlFor="category_hidden"
                    role="checkbox"
                    tabIndex="0"
                  >
                    <input
                      checked={false}
                      className="boolean optional switcher optional"
                      id="category_hidden"
                      name="hidden"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <div
                      className="checkbox"
                    />
                  </label>
                  <p
                    className="help-block"
                  >
                    Hidden categories only display to admins. These also display in reports
                  </p>
                </div>
              </div>
            </fieldset>
          </div>
          <footer
            className="confirm"
          >
            <button
              className="btn disabled"
              disabled={true}
              onClick={[Function]}
              type="submit"
            >
              Create Category
            </button>
            <button
              className="cancel close-modal"
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </footer>
        </form>
      </div>
    </section>
  </div>
</ForwardRef(Bootstrap(Modal))>
`;
